### DISCUSSION TITLE
Classification using local dataset (images)

### DISCUSSION THREAD

**Discussion by ggous** [azavea/raster-vision#2256] [General]
Hello, I am trying to accomplish a classification training.  I have tif images in various folders and a csv which refers to every image and what kind of label it has.

After successfully have done the training using :

```
ClassificationModelConfig
SolverConfig
ClassificationLearnerConfig
```
 

and a `CustomDataConfig`, something like:

```py
class CustomDataConfig(DataConfig):
    root_dir: Optional[str] = None
    
    def build(self, df, valid_split, test_split):
        # Shuffle the dataframe before splitting
        df = df.sample(frac=1).reset_index(drop=True)
       .....
      # Create datasets
        train_ds = CustomTIFDataset(
            df=df_train,
            root_dir=self.root_dir,
            transform=data_augmentation_transform,
        )

        valid_ds = CustomTIFDataset(
            df=df_valid,
            root_dir=self.root_dir,
            transform=None, 
        )
```

and 

```py
 learner = ClassificationLearner(

        cfg=learner_config,
        output_dir='data/output',
        train_ds=train_ds,
        valid_ds=valid_ds
    )


```
I want to proceed to predictions.

So, I am trying something like:

```py
predictions = learner.predict_dataset(test_ds,
                                      raw_out=True,          
                                      numpy_out=True,   
                                      progress_bar=True)

pred_labels = ChipClassificationLabels.from_predictions(
    test_ds,
    predictions)

scores = pred_labels.get_scores()
```

but the scores is an empty list.

I am not sure if I have to use `ChipClassificationLabels` though. I want a Classifications Labels, probably not Chip but I can't find something else!

**Comment by AdeelH** [comment]
You should be passing `test_ds.windows` to `ChipClassificationLabels.from_predictions()` and not `test_ds`.

```py
pred_labels = ChipClassificationLabels.from_predictions(
    test_ds.windows,
    predictions)
```

**Reply by ggous** [reply]
Hello, 

the thing is that I have tif images in folders and I just load them with rasterio. So, `train_ds`, `test_ds` do not have a `windows` method.

I tried to write one like:

```

    def get_windows(self):
        windows = []
        for idx in range(len(self)):
            image, _ = self[idx]
            window_size = (32, 32)
            for y in range(0, image.shape[0] - window_size[0] + 1, window_size[0]):
                for x in range(0, image.shape[1] - window_size[1] + 1, window_size[1]):
                    box = Box(x, y, x + window_size[1], y + window_size[0]) 
                    windows.append(box)
        return windows
```

My images have size (64, 64). Whatever size I am trying with  ` window_size = (32, 32)`, when i run `train_ds.get_windows()` it returns an empty list.

**Reply by AdeelH** [reply]
The prediction code above is designed to work with a `SlidingWindowGeoDataset` (as shown in [this tutorial](https://docs.rastervision.io/en/stable/usage/tutorials/pred_and_eval_ss.html#Get-scene-to-predict) for example). The idea is to make individual predictions for all the windows in the full image and then stitch them together using `Labels.from_predictions()`. If your images are already small and you do not want to split them further into smaller chips, you can set a size and stride (e.g. set both to 64) such that only one window is generated.

You can also skip `ChipClassificationLabels.from_predictions()` and just access the predictions directly. `predictions` in your code above is just an iterator, so you can do `list(predictions)` or loop over it to access the individual predictions and then save them however you like.

Hope this helps. Please provide more details if I have misunderstood what you're trying to do.

**Reply by ggous** [reply]
Thanks! I thought I have to use `ChipClassificationLabels`.
